### PR TITLE
[spirv] Add support vector swizzling

### PR DESCRIPTION
**Pull request by antiagainst**

### CODE REVIEW

File: tools/clang/lib/SPIRV/EmitSPIRVAction.cpp [hunk 1]
```diff
@@ -852,13 +974,193 @@ class SPIRVEmitter : public ASTConsumer {
...
+  /// Condenses a sequence of HLSLVectorElementExpr starting from the given
+  /// expr into one. Writes the original base into *basePtr and the condensed
+  /// accessor into *flattenedAccessor.
+  void condenseVectorElementExpr(
+      const HLSLVectorElementExpr *expr, const Expr **basePtr,
+      hlsl::VectorMemberAccessPositions *flattenedAccessor) {
+    llvm::SmallVector<hlsl::VectorMemberAccessPositions, 2> accessors;
+    accessors.push_back(expr->getEncodedElementAccess());
+
+    // Recursively dencending until we find the true base vector. In the
```
Comment (ehsannas) [inline tools/clang/lib/SPIRV/EmitSPIRVAction.cpp]: s/dencending/descending

File: tools/clang/lib/SPIRV/EmitSPIRVAction.cpp [hunk 2]
```diff
@@ -852,13 +974,193 @@ class SPIRVEmitter : public ASTConsumer {
...
+        return doExpr(baseExpr);
+      }
+
+      // If the base is an lvalue, we should emit an access chain instruction
+      // so that we can load/store the specified element. For rvalue base,
+      // we should use composite extraction. We should check the immediate base
+      // instead of the original base here since we can have something like
+      // v.xyyz to turn a lvalue v into rvalue.
+      if (expr->getBase()->isGLValue()) { // E.g., v.x;
+        // TODO: select the correct storage class
```
Comment (ehsannas) [inline tools/clang/lib/SPIRV/EmitSPIRVAction.cpp]: why not do it now? :) is it difficult to get the correct storage class? I'm afraid we may forget to do it! (file a case?)

File: tools/clang/lib/SPIRV/EmitSPIRVAction.cpp [hunk 3]
```diff
@@ -852,13 +974,193 @@ class SPIRVEmitter : public ASTConsumer {
...
+    const uint32_t baseVal = loadIfGLValue(baseExpr);
+    // Use base for both vectors. But we are only selecting values from the
+    // first one.
+    return theBuilder.createVectorShuffle(type, baseVal, baseVal, selectors);
+  }
+
+  /// Returns true if the given expression will be translated into a vector
+  /// shuffle instruction in SPIR-V.
+  ///
+  /// We emit a vector shuffle instruction iff selecting neither one elements
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/EmitSPIRVAction.cpp]: Done

File: tools/clang/lib/SPIRV/EmitSPIRVAction.cpp (hunk 1 above)
Comment (antiagainst) [inline tools/clang/lib/SPIRV/EmitSPIRVAction.cpp]: Done

File: tools/clang/lib/SPIRV/EmitSPIRVAction.cpp [hunk 4]
```diff
@@ -694,30 +716,132 @@ class SPIRVEmitter : public ASTConsumer {
...
+    // For example, for vec4.yz = vec2, we nee to do the following:
+    //
+    //   %vec4Val = OpLoad %v4float %vec4
+    //   %vec2Val = OpLoad %v2float %vec2
+    //   %shuffle = OpVectorShuffle %v4float %vec4Val %vec2Val 0 4 5 3
+    //   OpStore %vec4 %shuffle
+    //
+    // When doing the vector shuffle, we use the lhs base vector as the first
+    // vector and the rhs vector as the second vector. We select all elements
+    // in the second vector and all elements untouched in the first vector.
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/EmitSPIRVAction.cpp]: Sorry, misleading comment. Removed the second part.

File: tools/clang/lib/SPIRV/EmitSPIRVAction.cpp [hunk 5]
```diff
@@ -694,30 +716,132 @@ class SPIRVEmitter : public ASTConsumer {
...
+    const uint32_t baseTypeId = typeTranslator.translateType(baseType);
+    const uint32_t vec1 = doExpr(base);
+    const uint32_t vec1Val = theBuilder.createLoad(baseTypeId, vec1);
+    const uint32_t shuffle =
+        theBuilder.createVectorShuffle(baseTypeId, vec1Val, rhs, selectors);
+
+    theBuilder.createStore(vec1, shuffle);
+
+    // TODO: OK, this return value is incorrect for compound assignments, for
+    // which cases we should return lvalues. Should at least emit errors if
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/EmitSPIRVAction.cpp]: We are not handling many kind of variables now. This should be a thing to fix when adding support for global variables and more variables with semantics. #102 for this.

File: tools/clang/lib/SPIRV/EmitSPIRVAction.cpp (hunk 2 above)
Comment (antiagainst) [inline tools/clang/lib/SPIRV/EmitSPIRVAction.cpp]: We are not handling many kind of variables now. This should be a thing to fix when adding support for global variables and more variables with semantics. #102 for this.

**Pull request merged**